src/cls/queue/cls_queue_types.h

```diff
@@ -0,0 +1,90 @@
+#ifndef CEPH_CLS_QUEUE_TYPES_H
+#define CEPH_CLS_QUEUE_TYPES_H
+
+#include "include/types.h"
+
+#define QUEUE_HEAD_SIZE_1K 1024
+//Actual start offset of queue data
+#define QUEUE_START_OFFSET_1K QUEUE_HEAD_SIZE_1K
+
+#define QUEUE_HEAD_SIZE_4K (4 * 1024)
+//Actual start offset of queue data
+#define QUEUE_START_OFFSET_4K QUEUE_HEAD_SIZE_4K
+
+constexpr unsigned int QUEUE_HEAD_START = 0xDEAD;
+constexpr unsigned int QUEUE_ENTRY_START = 0xBEEF;
+
+struct cls_queue_marker
+{
+  uint64_t offset{0};
+  uint64_t gen{0};
+
+  void encode(bufferlist& bl) const {
+    ENCODE_START(1, 1, bl);
+    encode(gen, bl);
+    encode(offset, bl);
+    ENCODE_FINISH(bl);
+  }
+
+  void decode(bufferlist::const_iterator& bl) {
+    DECODE_START(1, bl);
+    decode(gen, bl);
+    decode(offset, bl);
+    DECODE_FINISH(bl);
+  }
+
+  string to_str() {
+    string marker = std::to_string(gen) + '/' + std::to_string(offset);
+    return marker;
+  }
+
+  void from_str(string& marker) {
```

Comment:
needs some error handling around stoull(). the c++ version std::stoull() throws exceptions on errors and needs a separate std::string as input. i'd recommend using the c strtoull() instead:
```c++
// decode the marker from a null-terminated string
int queue_marker::from_str(const char* str)
{
  errno = 0;
  char* end = nullptr;
  gen = ::strtoull(str, &end, 10);
  if (errno) {
    return errno;
  }
  if (str == end || *end != '/') { // expects delimiter
    return -EINVAL;
  }
  str = end + 1;
  offset = ::strtoull(str, &end, 10);
  if (errno) {
    return errno;
  }
  if (str == end || *end != 0) { // expects null terminator
    return -EINVAL;
  }
  return 0;
}
```